### DISCUSSION TITLE
Unifi Cloud Key Plus 2+

### DISCUSSION THREAD

**Discussion by nideville** [danielfernau/unifi-protect-video-downloader#45] [Q&A]
Hi,

I'm trying to use this project to retrieve video footage from my Unifi Cloud Key Plus running 2.0.27, Unifi protect 1.17.4
And I get a 404
![image](https://user-images.githubusercontent.com/65956408/114864412-424db480-9df1-11eb-9a9e-19c2fc2cce06.png)

I guess this is due to changes in Cloud Key OS version 2+
Any chance you already view and correct this ?

Best regards,
Nicolas

**Comment by danielfernau** [comment]
Hi Nicolas!

Yes, you're right: the newer UniFi Cloud Key Gen2+ firmware bundles install UniFi OS, which includes changes to the authentication logic and some parts of the API.

You might be able to connect to the Cloud Key and download files using the UDM Pro instructions --> https://github.com/unifi-toolbox/unifi-protect-video-downloader#for-udm-based-protect-installations

See also
https://github.com/unifi-toolbox/unifi-protect-video-downloader/issues/38
https://github.com/unifi-toolbox/unifi-protect-video-downloader/issues/25

I definitely know about these issues and what's causing them - but since there's a lot going on at my work for the past couple of months I don't have enough free time to actively work on this project. Hopefully things calm down a bit soon... once they do, I'll provide some updates that should fix the most important issues 🙂 

Best,
Daniel

**Reply by nideville** [reply]
ok got it. I manage to make the connection works (I get a "successfully authenticated" message) but I always get 0 file downloaded no matter what parameter I try. Any idea ? Is there a debug option ?

**Reply by danielfernau** [reply]
Without knowing the exact command you're using I can only point you to the _download_ command's documentation: https://github.com/unifi-toolbox/unifi-protect-video-downloader/wiki/Usage-(v2.x)#protect-archiver-download

In case you post your parameters here: please remove username/password/camera-IDs/other sensitive information!

There's no additional debug option available right now, as the program already prints a lot of the relevant information to the console anyways. Make sure you select the correct camera(s), and ensure that they have recordings available within the time span you specify.

**Reply by nideville** [reply]
Here is the command I run (after running the proxy):
`docker run --network=unifi-protect-video-downloader_default --volume /Users/ndeville/Documents/99_PERSO/video:/downloads unifitoolbox/protect-archiver download --address="unifi-udm-api-proxy" --username XXXXX --password XXXXXX --start "2021-04-13 08:00:00" --end "2021-04-13 16:00:00" --touch-files --skip-existing-files  /downloads`
I'm sure they are some recording for this day and I have only 1 camera so I leave the default parameter (all)

And I get this output:
```
Getting camera list
Successfully authenticated as user local
0 files downloaded (0.0 b), 0 files skipped, 0 files total
```

**Reply by danielfernau** [reply]
Looks like the program isn't even trying to download anything. Just a guess, but you could try a slightly different format for the arguments like this: `docker run --network=unifi-protect-video-downloader_default --volume="/Users/ndeville/Documents/99_PERSO/video:/downloads" unifitoolbox/protect-archiver download --address="unifi-udm-api-proxy" --username="XXXXX" --password="XXXXXX" --start="2021-04-13T08:00:00+0200" --end="2021-04-13T16:00:00+0200" --touch-files --skip-existing-files /downloads`
_(I've changed the path, username, password and start/end times format in the example above.)_
If that doesn't help, you could also try explicitly adding `--cameras=all` to manually select all available cameras.
Keep in mind that the camera(s) need(s) to be online and connected to the CloudKey while trying to download content using the _all_ option. If they're not connected, their content will not be downloaded. It's a "known issue".

**Reply by nideville** [reply]
well same result ... (the start/end options were not allowed, I had to remove the +XXXX).
Anyway, I will try to clone the git repo and play with the python script to try to find out what is the issue :)

**Reply by nideville** [reply]
I found the issue but not how to solve it :)
The` /api/cameras` route answers with a 500 Internal Server Error ..... maybe an issue with the proxy

**Reply by danielfernau** [reply]
Thanks for the update! Yes, in that case it's probably a proxy issue. Well, not directly, since it only maps/translates the requests to different paths. So it's a combination of (1) a request from the application and (2) the translation done by the proxy, which the CloudKey doesn't quite know how to process, resulting in a status 500. Since the API is not officially documented, fixing the issue requires a bit of R&D work. I'll post an update if I find anything. In case you want to have a look at the API yourself: you can use the Network tab of your web browser's Developer Console while navigating through the CloudKey's web interface. You can see the URLs, requests and responses there.

**Reply by nideville** [reply]
Ok I will have a look ! I posted an issue on the proxy project and I get this answer : "the proxy is not needed anymore". I'm not sure what it means :) https://github.com/Silvenga/unifi-udm-api-proxy/issues/14

**Reply by danielfernau** [reply]
Well, in that case the API might have been updated by UI to restore the "old" paths that were available before UniFiOS was a thing. That would mean that this should work then: `docker run --volume="/Users/ndeville/Documents/99_PERSO/video:/downloads" unifitoolbox/protect-archiver download --address="[[YOUR-CLOUDKEY-IP]]" --port=443 --username="XXXXX" --password="XXXXXX" --start "2021-04-13 08:00:00" --end "2021-04-13 16:00:00" --touch-files --skip-existing-files /downloads` -- it's either port 443 or 7443.
Let me know if that works. If it does, I'll go ahead and update the README accordingly.

**Reply by nideville** [reply]
that was my 1st try after briis answer :) but I get a 404 on the auth

**Reply by danielfernau** [reply]
Based on that I'd think that authentication and Protect API access are on different ports, i.e. Auth on 443 and Protect on 7443 or something like that. Either way, a re-write of the authentication part of the software is overdue and once I get to it, it will probably solve most of the open issues at once.

I'll have a look as soon as possible.
(I know, I keep saying that – in the comments of other open issues as well – but my job leaves little to no time for any personal projects like this one at the end of the day, and it's been like that for quite a while now. That's why updates take so long. Sorry.)

**Reply by nideville** [reply]
I finally succeed in modifying your script to work with my UCK.
Everything in on port 443 no need for API-PROXY

**Auth**

Auth route is now on `/api/auth/login`
In return you get the TOKEN in Set-Cookie header ==> `authorization_header = response.cookies.get("TOKEN")`
Every time you need to use the token you need to put it in the cookies ==> `cookies={'TOKEN' : self.get_api_token()},`

**Get cameras list**
The new route is : `/proxy/protect/api/cameras`

**Get footage**
the new route is : `/proxy/protect/api/video/export?&camera={camera.id}&start={js_timestamp_range_start}&end={js_timestamp_range_end}`


I guess there are many other changes to update your scripts :) but for my basic usage it's enough !

**Reply by danielfernau** [reply]
Hey, great work! Thanks a lot for sharing your progress, that definitely speeds things up with fixing the auth problems. I'll remove the proxy and replace the relevant parts of the software then; should be able to do that over the weekend.

**Reply by danielfernau** [reply]
Hi @nideville – managed to implement the changes yesterday and today. The new release – [v2.0.1](https://github.com/unifi-toolbox/unifi-protect-video-downloader/releases/tag/v2.0.1) – is now available. Please have a look, it should now be fully compatible with UniFi OS-based Protect systems. Let me know in case you encounter any new (or old) problems. I've also removed the UDM API Proxy from code and documentation as it is indeed no longer required.
